### ISSUE TITLE
normalize_quaternion returns all-NaN in float16 for the zero quaternion: the default eps=1e-12 underflows to 0

### ISSUE THREAD

**Issue by ducha-aiki** [kornia/kornia#4021]
### Describe the bug

`normalize_quaternion` returns `[nan, nan, nan, nan]` in `float16` for the zero quaternion. `float32`, `float64` and `bfloat16` all return `[0, 0, 0, 0]` for the same input.

The cause is a dtype-dependent underflow of the **default** `eps`, not the quaternion maths. The implementation is

```python
def normalize_quaternion(quaternion: torch.Tensor, eps: float = 1.0e-12) -> torch.Tensor:
    ...
    return F.normalize(quaternion, p=2.0, dim=-1, eps=eps)
```

(`kornia/geometry/conversions.py`). `F.normalize` divides by `max(||q||, eps)`, but `float16`'s smallest subnormal is ≈ `6.1e-5`, so the default `1e-12` rounds to exactly `0.0` and the clamp becomes a no-op. The division is then `0 / 0 → NaN`.

`bfloat16` is unaffected because its exponent range is `float32`'s, so `1e-12` is representable there.

This is the same underflow class as #3966 (`quaternion_exp_to_log`, default `eps=1e-8`, fixed in #3990) — a different function with the same root cause, so fixing one did nothing for the other.

**It is not the same defect as #3952.** #3952 is about the clamp *working* and producing a non-unit result (`[1e-13, 0, 0, 0] -> [0.1, 0, 0, 0]`, a `||q|| / eps` scaling). This issue is about the clamp *not existing* at `float16`, so there is no guard at all. #3952's own repro output records `normalize_quaternion(zeros(4)) == [0, 0, 0, 0]`, which is untrue at `float16`; the two need separate fixes and a fix for either leaves the other standing.

### Reproduction steps

```python
import torch
from kornia.geometry.conversions import normalize_quaternion, quaternion_to_rotation_matrix

for dtype in (torch.float64, torch.float32, torch.bfloat16, torch.float16):
    q = torch.zeros(4, dtype=dtype)
    print(dtype, normalize_quaternion(q).tolist())
```

```
torch.float64  [0.0, 0.0, 0.0, 0.0]
torch.float32  [0.0, 0.0, 0.0, 0.0]
torch.bfloat16 [0.0, 0.0, 0.0, 0.0]
torch.float16  [nan, nan, nan, nan]
```

The eps underflow is the whole story — passing an eps that survives `float16` fixes it:

```python
>>> normalize_quaternion(torch.zeros(4, dtype=torch.float16), eps=6.104e-05).tolist()
[0.0, 0.0, 0.0, 0.0]
>>> float(torch.tensor(1e-12, dtype=torch.float16))    # 0.0      <- the default underflows
0.0
>>> float(torch.tensor(1e-12, dtype=torch.bfloat16))   # representable
1.0018652574217413e-12
```

A non-zero quaternion is fine at the same dtype, which is why the suite has never caught it:

```python
>>> normalize_quaternion(torch.tensor([1., 0., 1., 0.], dtype=torch.float16)).tolist()
[0.70703125, 0.0, 0.70703125, 0.0]
```

### Blast radius

Both in-tree callers inherit it.

`quaternion_to_rotation_matrix` calls `normalize_quaternion` first, so the zero quaternion becomes an all-NaN matrix at `float16` where every other dtype returns the identity:

```python
for dtype in (torch.float64, torch.float32, torch.bfloat16, torch.float16):
    print(dtype, quaternion_to_rotation_matrix(torch.zeros(4, dtype=dtype)).flatten().tolist())
```

```
torch.float64  [1.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 1.0]
torch.float32  [1.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 1.0]
torch.bfloat16 [1.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 1.0]
torch.float16  [nan, nan, nan, nan, nan, nan, nan, nan, nan]
```

`Quaternion.normalize()` reaches it through the same call (`kornia/geometry/quaternion.py:609`):

```python
>>> Quaternion(torch.zeros(1, 4, dtype=torch.float16)).normalize().data.tolist()
[[nan, nan, nan, nan]]
>>> Quaternion(torch.zeros(1, 4, dtype=torch.float32)).normalize().data.tolist()
[[0.0, 0.0, 0.0, 0.0]]
```

### Expected behavior

`float16` should behave like the other three dtypes: the guard should hold at every dtype the function accepts, so the zero quaternion returns `[0, 0, 0, 0]` rather than NaN. Whatever the eventual answer to #3952 is (return a unit quaternion, raise, or keep passing the zero through), it should be the *same* answer in `float16` as in `float64`, which today it is not.

### Notes for a fix

The shape used for #3966 is a dtype-aware floor. `max(eps, torch.finfo(dtype).tiny)` works — `finfo.tiny` is the right choice over `finfo.eps` for a division guard, since `float16`'s are `6.1e-5` and `9.8e-4` respectively and the larger one would distort ordinary sub-`1e-3` inputs.

One catch a fix has to handle deliberately: `normalize_quaternion` takes `eps` as a public argument, and `eps=0.0` currently means "no guard, let it divide by zero" — pinned by `test_wart_normalize_quaternion_without_eps_divides_by_zero_3952` in `tests/geometry/test_conversions.py`. A plain `max(eps, finfo.tiny)` silently overrides an explicit `eps=0` and trips that pin at `float32`/`float64`. Either the floor applies only to the default, or the pin's contract changes on purpose.

The defect is already recorded in the suite as a `pytest.skip` inside `test_wart_zero_quaternion_becomes_the_identity_matrix_3952`, which skips `float16` with this exact explanation. That skip is what should be removed when this is fixed.

Also note #3954, which is live in the same neighbourhood: `quaternion_to_rotation_matrix` upcasts unbatched half-precision input to `float32`. The `float16` cells above are unbatched, so a fix for #3954 will change what dtype those NaNs come back in without making them stop being NaN.

## Disposition

**n/a — single symbol.** `normalize_quaternion` has no near-duplicate in the ledger; this is a defect report, not a duplication finding.

### Environment

Reproduced on `main` at `4da60edc`, CPU, torch 2.9.1, Python 3.11.14, kornia 0.9.0rc1. Not device-specific — it is a property of the `float16` exponent range, not of a backend.
